src/operator/bilinear_sampler.cc

```diff
@@ -0,0 +1,171 @@
+              top_left_v = *(data + data_index);
+            }
+            if (between(top_left_x+1, 0, i_w-1) && between(top_left_y, 0, i_h-1)) {
+              *(g_input + data_index + 1) += *(grad + grad_index) * top_left_y_w
+                                              * (1.0 - top_left_x_w);
+              top_right_v = *(data + data_index + 1);
+            }
+            if (between(top_left_x, 0, i_w-1) && between(top_left_y+1, 0, i_h-1)) {
+              *(g_input + data_index+ i_w) += *(grad + grad_index) * (1.0 - top_left_y_w)
+                                              * top_left_x_w;
+              bottom_left_v = *(data + data_index + i_w);
+            }
+            if (between(top_left_x+1, 0, i_w-1) && between(top_left_y+1, 0, i_h-1)) {
+              *(g_input + data_index+ i_w + 1) += *(grad + grad_index) * (1.0 - top_left_y_w)
+                                                  * (1.0 - top_left_x_w);
+              bottom_right_v = *(data + data_index + i_w + 1);
+            }
+            // calc weight grad of top_left_w, then multiple -1 is the grad of grid_src
+            top_left_y_gw -= *(grad + grad_index) * (top_right_v - bottom_right_v +
+                              (top_left_v - top_right_v - bottom_left_v + bottom_right_v)
+                              * top_left_x_w);
+            top_left_x_gw -= *(grad + grad_index) * (bottom_left_v - bottom_right_v +
+                              (top_left_v - top_right_v - bottom_left_v + bottom_right_v)
+                              * top_left_y_w);
+          }
+          // calc grad of grid
+          *(grad_grid + grid_src_index + o_h * o_w) += top_left_y_gw * (i_h - 1) / 2;
+          *(grad_grid + grid_src_index) += top_left_x_gw * (i_w - 1) / 2;
+        }
+      }
+    }
+  }
+}  // namespace mshadow
+
+namespace mxnet {
+namespace op {
+template<>
+Operator* CreateOp<cpu>(BilinearSamplerParam param, int dtype) {
+  Operator *op = NULL;
+  MSHADOW_REAL_TYPE_SWITCH(dtype, DType, {
+    op = new BilinearSamplerOp<cpu, DType>(param);
+  })
+  return op;
+}
+
+Operator *BilinearSamplerProp::CreateOperatorEx(Context ctx, std::vector<TShape> *in_shape,
+                                     std::vector<int> *in_type) const {
+  std::vector<TShape> out_shape, aux_shape;
+  std::vector<int> out_type, aux_type;
+  CHECK(InferType(in_type, &out_type, &aux_type));
+  CHECK(InferShape(in_shape, &out_shape, &aux_shape));
+  DO_BIND_DISPATCH(CreateOp, param_, (*in_type)[0]);
+}
+
+DMLC_REGISTER_PARAMETER(BilinearSamplerParam);
+
+MXNET_REGISTER_OP_PROPERTY(BilinearSampler, BilinearSamplerProp)
+.add_argument("data", "Symbol", "Input data to the BilinearsamplerOp.")
+.add_argument("grid", "Symbol", "Input grid to the BilinearsamplerOp."
+                                "grid has two channels: x_src, y_src")
+.add_arguments(BilinearSamplerParam::__FIELDS__())
+.describe("Apply bilinear sampling to input feature map.\n    "

```

Comment:
I still think some people will search for spatial transformer when looking for this layer.
Let's say "bilinear sampling (spatial transformer layer)" here and quote some relevent papers by title